### ISSUE TITLE
[rancher-dns] Rancher DNS record available before service container is ready

### ISSUE THREAD

**Issue by bdentino** [rancher/rancher#5792]
**Rancher Version:**
v1.1.2
**Docker Version:**
1.11.2
**OS and where are the hosts located? (cloud, bare metal, etc):**
RancherOS 4.4.10 - AWS
**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node, external DB
**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle
**Steps to Reproduce:**
A little work to set up, but so far reliably reproducible:
1. SSH into an existing rancher host
2. Start a docker container with rancher managed networking (I'm using alpine:3.4 with drill installed). In this container, monitor the dns records for one of your services: `watch -n1 drill <service>.<stack>.rancher.internal`
3. Force the monitored service to deploy onto a host that does not already have the image. You should notice that a dns record is added for the new container while the image is still being pulled. 
**Results:** 
A DNS record for the pending container is created immediately. I notice that once the container starts, the dns record is removed until it is healthy. But it's there while the image is pulling. When pulling large images or on hosts deploying a lot of containers simultaneously, this bad record can cause substantial internal networking problems for 5-10 minutes. For added effect, run another container to periodically curl your service using rancher's dns. Some of the requests will likely hit the bad record and time out. 
**Expected:**
The DNS record shouldn't be made available at least until the container is running, preferably not until after it is passing health checks, if any.

**Comment by alena1108** [comment]
@bdentino should be fixed with https://github.com/rancher/rancher-dns/pull/35/commits/96e9c4e73b12ba6fb67a88c50cf01b14b2a54483 - refactored dns microservice. 

Note that when service has 0 healthy records, it will be resolved to one of the unhealthy records. So service fqdn always resolves to something But when service has at least one healthy record, all the unhealthy records will be eliminated.

**Comment by sangeethah** [comment]
Tested with latest build from master:

rancher/dns:v0.10.4
rancher/healthcheck:v0.1.0
rancher/net:v0.7.0
rancher/metadata:v0.6.3
rancher/plugin-manager:v0.2.1
rancher/scheduler:v0.1.2


DNS for service still returns ip address of the containers that still in "starting" state when pulling images.

dig on service name returns - "10.42.78.182" ip address for the container which is still pulling images.

```
mysql> select id,state,health_state,name from instance where name like "%globalservice-health%" and removed is null;
+-----+----------+--------------+--------------------------------+
| id  | state    | health_state | name                           |
+-----+----------+--------------+--------------------------------+
|  85 | running  | healthy      | Default-globalservice-health-2 |
|  96 | running  | healthy      | Default-globalservice-health-1 |
| 130 | running  | healthy      | Default-globalservice-health-3 |
| 159 | running  | healthy      | Default-globalservice-health-5 |
| 163 | running  | healthy      | Default-globalservice-health-4 |
| 388 | starting | NULL         | Default-globalservice-health-6 |
+-----+----------+--------------+--------------------------------+
6 rows in set (0.00 sec)

mysql> select * from nic where instance_id=388;
+-----+------+------------+------+--------------------------------------+-------------+---------+---------------------+---------------------+---------------------+------+-------------+------------+-----------+---------+---------------+-------------------+
| id  | name | account_id | kind | uuid                                 | description | state   | created             | removed             | remove_time         | data | instance_id | network_id | subnet_id | vnet_id | device_number | mac_address       |
+-----+------+------------+------+--------------------------------------+-------------+---------+---------------------+---------------------+---------------------+------+-------------+------------+-----------+---------+---------------+-------------------+
| 388 | NULL |          5 | nic  | 1510758c-7fe3-4a5d-a07e-e15a094a3341 | NULL        | removed | 2016-11-16 21:16:07 | 2016-11-16 21:17:07 | 2016-11-16 21:26:05 | {}   |         388 |          5 |      NULL |    NULL |             0 | 02:b4:c5:e3:40:f6 |
+-----+------+------------+------+--------------------------------------+-------------+---------+---------------------+---------------------+---------------------+------+-------------+------------+-----------+---------+---------------+-------------------+
1 row in set (0.00 sec)

mysql> select * from ip_address_nic_map where nic_id=388;
+-----+------+-----------------+--------------------------------------+-------------+--------+---------------------+---------+-------------+------+---------------+--------+
| id  | name | kind            | uuid                                 | description | state  | created             | removed | remove_time | data | ip_address_id | nic_id |
+-----+------+-----------------+--------------------------------------+-------------+--------+---------------------+---------+-------------+------+---------------+--------+
| 316 | NULL | ipAddressNicMap | df8494f0-5f1c-4926-938f-81b4fad4623c | NULL        | active | 2016-11-16 21:16:08 | NULL    | NULL        | {}   |           326 |    388 |
+-----+------+-----------------+--------------------------------------+-------------+--------+---------------------+---------+-------------+------+---------------+--------+
1 row in set (0.00 sec)

mysql> select * from ip_address where id=326;
+-----+--------------+------------+-----------+--------------------------------------+-------------+----------+---------------------+---------+-------------+------+--------------+-----------+------------+-----------+---------+----------+------------+
| id  | name         | account_id | kind      | uuid                                 | description | state    | created             | removed | remove_time | data | address      | subnet_id | network_id | is_public | role    | hostname | ip_pool_id |
+-----+--------------+------------+-----------+--------------------------------------+-------------+----------+---------------------+---------+-------------+------+--------------+-----------+------------+-----------+---------+----------+------------+
| 326 | 10.42.78.182 |          5 | ipAddress | ec6ae015-b3ef-4a17-9119-59b49a0c1891 | NULL        | inactive | 2016-11-16 21:16:08 | NULL    | NULL        | {}   | 10.42.78.182 |         1 |          5 |           | primary | NULL     |       NULL |
+-----+--------------+------------+-----------+--------------------------------------+-------------+----------+---------------------+---------+-------------+------+--------------+-----------+------------+-----------+---------+----------+------------+
1 row in set (0.00 sec)
```

**Comment by alena1108** [comment]
@sangeethah fix should be in v0.10.6

**Comment by sangeethah** [comment]
Tested with v1.2.0-pre4-rc9 and rancher/dns:v0.10.6
DNS for service do not returns ip address of the containers that still in "starting" state when pulling images.
